AsyncDisplayKit/ASRunLoopQueue.mm

```diff
@@ -0,0 +1,98 @@
+//
+//  ASRunLoopQueue.m
+//  AsyncDisplayKit
+//
+//  Created by Rahul Malik on 3/7/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASRunLoopQueue.h"
+#import "ASThread.h"
+
+#import <deque>
+
+@interface ASRunLoopQueue () {
+  CFRunLoopRef _runLoop;
+  CFRunLoopObserverRef _runLoopObserver;
+  std::deque<id> __renderQueue;
+  ASDN::RecursiveMutex __renderQueueLock;
+}
+@property (nonatomic, copy) void (^queueConsumer)(id dequeuedItem, BOOL isQueueDrained);
+
+@end
+
+@implementation ASRunLoopQueue
+
+- (instancetype)initWithRunLoop:(CFRunLoopRef)runloop andHandler:(void(^)(id dequeuedItem, BOOL isQueueDrained))handlerBlock
+{
+  if (self = [super init]) {
+    _runLoop = runloop;
+    __renderQueue = std::deque<id>();
+    _queueConsumer = [handlerBlock copy];
+    void (^handlerBlock) (CFRunLoopObserverRef observer, CFRunLoopActivity activity) = ^(CFRunLoopObserverRef observer, CFRunLoopActivity activity) {
+      [self processQueue];
+    };
+    _runLoopObserver = CFRunLoopObserverCreateWithHandler(NULL, kCFRunLoopBeforeWaiting, true, 0, handlerBlock);
+    CFRunLoopAddObserver(_runLoop,
+                         _runLoopObserver,
+                         kCFRunLoopCommonModes);
+
+  }
+  return self;
+}
+
+- (void)processQueue
+{
+  std::deque<id> displayingNodes = std::deque<id>();
+  BOOL isQueueDrained = NO;
+  CFAbsoluteTime timestamp = 0;
+  {
+    ASDN::MutexLocker l(__renderQueueLock);
+
+    // Early-exit if we don't have any nodes to render.
+    if (__renderQueue.empty()) {
+      return;
+    }
+
+    // Snatch the next batch of nodes.
+    NSUInteger totalNodeCount = __renderQueue.size();
+    for (int i = 0; i < MIN(self.batchSize, totalNodeCount); i++) {
+      id node = __renderQueue[0];
+      displayingNodes.push_back(node);
+      __renderQueue.pop_front();
+    }
+
+    if (__renderQueue.empty()) {
+      isQueueDrained = YES;
+      timestamp = CFAbsoluteTimeGetCurrent();
+    }
+  }
+
+  for (int i = 0; i < displayingNodes.size(); i++) {
+    if (isQueueDrained && i == displayingNodes.size() - 1) {
```

Comment:
This logic is not immediately obvious to me, for example the hardcoded YES below, in a future update add a comment or two in this area